ClassCastException when using OpenTelemetry and max-connections property

### Describe the bug

Using OpenTelemetry and setting **quarkus.http.limits.max-connection** property results in ClassCastException 


### Expected behavior

No error

### Actual behavior

```
2024-10-14 11:06:26,807 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to recover after failed start: java.lang.RuntimeException: java.lang.RuntimeException: Unable to start HTTP server
        at io.quarkus.vertx.http.runtime.VertxHttpRecorder.startServerAfterFailedStart(VertxHttpRecorder.java:307)
        at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup.handleFailedInitialStart(VertxHttpHotReplacementSetup.java:73)
        at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.startupFailed(RuntimeUpdatesProcessor.java:1299)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:128)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:430)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:57)
        at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:138)
        at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:93)
        at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:131)
        at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:62)
Caused by: java.lang.RuntimeException: Unable to start HTTP server
        at io.quarkus.vertx.http.runtime.VertxHttpRecorder.doServerStart(VertxHttpRecorder.java:947)
        at io.quarkus.vertx.http.runtime.VertxHttpRecorder.startServerAfterFailedStart(VertxHttpRecorder.java:299)
        ... 9 more
Caused by: java.util.concurrent.ExecutionException: java.lang.ClassCastException: class io.quarkus.opentelemetry.runtime.tracing.intrumentation.vertx.OpenTelemetryVertxMetricsFactory$1 cannot be cast to class io.quarkus.vertx.http.r
untime.ExtendedQuarkusVertxHttpMetrics (io.quarkus.opentelemetry.runtime.tracing.intrumentation.vertx.OpenTelemetryVertxMetricsFactory$1 and io.quarkus.vertx.http.runtime.ExtendedQuarkusVertxHttpMetrics are in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @5ad851c9)
        at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
        at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
        at io.quarkus.vertx.http.runtime.VertxHttpRecorder.doServerStart(VertxHttpRecorder.java:864)
        ... 10 more
Caused by: java.lang.ClassCastException: class io.quarkus.opentelemetry.runtime.tracing.intrumentation.vertx.OpenTelemetryVertxMetricsFactory$1 cannot be cast to class io.quarkus.vertx.http.runtime.ExtendedQuarkusVertxHttpMetrics (i
o.quarkus.opentelemetry.runtime.tracing.intrumentation.vertx.OpenTelemetryVertxMetricsFactory$1 and io.quarkus.vertx.http.runtime.ExtendedQuarkusVertxHttpMetrics are in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @5ad851c9)
        at io.quarkus.vertx.http.runtime.VertxHttpRecorder$WebDeploymentVerticle.setupTcpHttpServer(VertxHttpRecorder.java:1278)
        at io.quarkus.vertx.http.runtime.VertxHttpRecorder$WebDeploymentVerticle.start(VertxHttpRecorder.java:1222)
        at io.vertx.core.impl.DeploymentManager.lambda$doDeploy$5(DeploymentManager.java:210)
        at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:270)
        at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:252)
        at io.vertx.core.impl.ContextInternal.lambda$runOnContext$0(ContextInternal.java:50)
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:994)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:840)


```

### How to Reproduce?

Follow the instructions from https://quarkus.io/guides/opentelemetry-metrics:
1. quarkus create app org.acme:opentelemetry-quickstart \
    --extension='rest,quarkus-opentelemetry' \
    --no-code
2. add quarkus.http.limits.max-connections=50 property in application.properties
3. run application: quarkus dev

### Output of `uname -a` or `ver`

Microsoft Windows [Version 10.0.19045.4529]

### Output of `java -version`

temurin-21.0.3

### Quarkus version or git rev

3.15.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)

### Additional information

_No response_

--- a/extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/metrics/ClassCastExceptionTest.java
+++ b/extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/metrics/ClassCastExceptionTest.java
@@ -1,0 +1,44 @@
+package io.quarkus.opentelemetry.deployment.metrics;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.opentelemetry.deployment.common.exporter.InMemoryMetricExporter;
+import io.quarkus.opentelemetry.deployment.common.exporter.InMemoryMetricExporterProvider;
+import io.quarkus.opentelemetry.deployment.common.exporter.TestSpanExporter;
+import io.quarkus.opentelemetry.deployment.common.exporter.TestSpanExporterProvider;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class ClassCastExceptionTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest TEST = new QuarkusUnitTest()
+            .setArchiveProducer(
+                    () -> ShrinkWrap.create(JavaArchive.class)
+                            .addClasses(TestSpanExporter.class, TestSpanExporterProvider.class)
+                            .addClasses(InMemoryMetricExporter.class, InMemoryMetricExporterProvider.class)
+                            .addAsResource(new StringAsset(InMemoryMetricExporterProvider.class.getCanonicalName()),
+                                    "META-INF/services/io.opentelemetry.sdk.autoconfigure.spi.metrics.ConfigurableMetricExporterProvider")
+                            .add(new StringAsset(
+                                    "quarkus.otel.metrics.enabled=true\n" +
+                                            "quarkus.otel.metrics.exporter=in-memory\n" +
+                                            "quarkus.otel.metric.export.interval=300ms\n" +
+                                            "quarkus.http.limits.max-connections=50\n" +
+                                            "quarkus.otel.traces.exporter=none\n" +
+                                            "quarkus.otel.logs.exporter=none\n"),
+                                    "application.properties"));
+
+    /**
+     * ClassCastException when using OpenTelemetry and max-connections property
+     * See https://github.com/quarkusio/quarkus/issues/43852
+     */
+    @Test
+    public void testReproducer() {
+        // This test is successful if it does not throw an exception
+        Assertions.assertTrue(true);
+    }
+}

--- a/extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/intrumentation/vertx/OpenTelemetryVertxMetricsFactory.java
+++ b/extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/intrumentation/vertx/OpenTelemetryVertxMetricsFactory.java
@@ -2,6 +2,7 @@ package io.quarkus.opentelemetry.runtime.tracing.intrumentation.vertx;
 
 import java.util.Optional;
 
+import io.quarkus.vertx.http.runtime.ExtendedQuarkusVertxHttpMetrics;
 import io.vertx.core.Context;
 import io.vertx.core.VertxOptions;
 import io.vertx.core.http.HttpServerOptions;
@@ -16,7 +17,7 @@ import io.vertx.core.spi.observability.HttpRequest;
  * This is used to retrieve the route name from Vert.x. This is useful for OpenTelemetry to generate the Span name and
  * <code>http.route</code> attribute. Right now, there is no other way to retrieve the route name from Vert.x using the
  * Telemetry SPI, so we need to rely on the Metrics SPI.
- *
+ * <p>
  * Right now, it is not possible to register multiple <code>VertxMetrics</code>, meaning that only a single one is
  * available per Quarkus instance. To avoid clashing with other extensions that provide Metrics data (like the
  * Micrometer extension), we only register the {@link OpenTelemetryVertxMetricsFactory} if the
@@ -25,17 +26,19 @@ import io.vertx.core.spi.observability.HttpRequest;
 public class OpenTelemetryVertxMetricsFactory implements VertxMetricsFactory {
     @Override
     public VertxMetrics metrics(final VertxOptions options) {
-        return new VertxMetrics() {
-            @Override
-            public HttpServerMetrics<?, ?, ?> createHttpServerMetrics(final HttpServerOptions options,
-                    final SocketAddress localAddress) {
-                return new OpenTelemetryHttpServerMetrics();
-            }
-        };
+        return new OpenTelemetryHttpServerMetrics();
     }
 
     public static class OpenTelemetryHttpServerMetrics
-            implements HttpServerMetrics<OpenTelemetryHttpServerMetrics.MetricRequest, Object, Object> {
+            implements HttpServerMetrics<OpenTelemetryHttpServerMetrics.MetricRequest, Object, Object>,
+            VertxMetrics, ExtendedQuarkusVertxHttpMetrics {
+
+        @Override
+        public HttpServerMetrics<?, ?, ?> createHttpServerMetrics(final HttpServerOptions options,
+                final SocketAddress localAddress) {
+            return this;
+        }
+
         @Override
         public MetricRequest requestBegin(final Object socketMetric, final HttpRequest request) {
             return MetricRequest.request(request);
@@ -46,6 +49,12 @@ public class OpenTelemetryVertxMetricsFactory implements VertxMetricsFactory {
             if (route != null) {
                 requestMetric.getContext().ifPresent(context -> context.putLocal("VertxRoute", route));
             }
+        }
+
+        @Override
+        public ConnectionTracker getHttpConnectionTracker() {
+            // To be implemented if we decide to instrument with OpenTelemetry. See VertxMeterBinderAdapter for an example.
+            return ExtendedQuarkusVertxHttpMetrics.NOOP_CONNECTION_TRACKER;
         }
 
         static final class MetricRequest {

